### DISCUSSION TITLE
New layer containing an AlpineJS component gets initialize 5 times

### DISCUSSION THREAD

**Discussion by johndwells** [unpoly/unpoly#314] [Q&A] [answered]
I'm building an app that loads a form into an Unpoly modal drawer. This form is an AlpineJS component. I am finding that the component is being initialized **five** times. If I alternatively load the form into the root layer, it only initializes once.

I've tried a simpler component, a mere empty alpine component with only a `console.log()` on its `init()` method, and again the same happens - if it's rendered in a modal layer, it logs **five** messages to the console; if it's rendered at the root layer, only once.

I've been using Unpoly & Alpine together without issue for some time, but this is admittedly the first time I've tried to render an Alpine component into a non-root layer.

I would really appreciate any insight into this matter... :pray:

Unpoly Version: 2.4.1
AlpineJS Version: 3.5.1

Edit: Here's a Glitch project demonstration behavior: https://glitch.com/edit/#!/unpoly-discussion-314?path=index.html%3A33%3A0

**Comment by adam12** [comment]
Could you make a repro using the [Glitch starter project](https://glitch.com/edit/#!/empty-unpoly-project)?

**Reply by johndwells** [reply]
@adam12 Absolutely - just updated the post with a link.

**Reply by adam12** [reply]
Demo helped, but I'm not sure there's going to be a great solution here. I'll wait to see if anybody else has any other ideas.

**Comment by triskweline** [comment, marked as answer]
I think there could be an issue in Alpine's [`MutationObserver`](https://github.com/alpinejs/alpine/blob/main/packages/alpinejs/src/mutation.js) that evaluates `[x-init]` when a new element is attached to the DOM. The issue can be observed when the new element with `[x-init]` is not attached directly, but is a *child* of the attached element.

To give a minimal example without Unpoly, the following code appends a single element with `[x-init]`, but the callback in `[x-init]` is called twice:

```js
container = document.createElement('div')
document.body.appendChild(container)
alpineElement = document.createElement('span')
alpineElement.setAttribute('x-data', '')
alpineElement.setAttribute('x-init', 'console.log("alpine init")')
container.appendChild(alpineElement) // prints "alpine init" twice
```

**Reply by johndwells** [reply]
@triskweline This is quite interesting. I can take this to the Alpine github discussions and see what they think...

But still - what could be different about inserting fragments into the _root_ layer, versus a modal or similar? Like I said I've been combining AlpineJS & Unpoly for a while now, and I've never had an issue. Sometimes my root layer has been the `<body>` element by default, but usually it's some child `<div>`.

**Reply by johndwells** [reply]
@triskweline After sleeping on it I think I see what you're saying - your snippet encapsulates the difference between rendering on the root - e.g. an element that's **already** in the DOM - and a modal, which is being _injected_ into the DOM.

This now explains why in my experimentation, modals & drawers all resulted in my init code running **five** times, whereas a _popup_ only resulted in it running **three** - because the number of elements used to create the overlays: https://unpoly.com/customizing-overlays 💡

Also, I can modify your code to avoid the duplicate log, but wrapping the child element code in a `setTimeout()`:

```
container = document.createElement('div')
document.body.appendChild(container)
setTimeout(() => {
  alpineElement = document.createElement('span')
  alpineElement.setAttribute('x-data', '')
  alpineElement.setAttribute('x-init', 'console.log("alpine init")')
  container.appendChild(alpineElement) // prints "alpine init" once
});

```

My understanding is that `setTimeout()` lets the browser finish actually updating the DOM, before running our callback (I've listened to Caleb, the author behind AlpineJS, talk about this concept on one of his podcasts).

This is super helpful, thank you! I'm going to raise this in the Alpine discussion and see what comes of it.

**Reply by johndwells** [reply]
Discussion in Alpine's community started: https://github.com/alpinejs/alpine/discussions/2372

**Reply by johndwells** [reply]
The bug in Alpine has been confirmed, and a PR for the fix underway. Thank you so much @triskweline for your wise insight into this one!

**Reply by triskweline** [reply]
Good news :slightly_smiling_face:
